# IgG+ Specimens - Lineage analysis with SONAR

As I've implemented it, SONAR uses pRESTO's merged, quality-filtered,
primer-matching reads as input.

## SONAR Analysis - Module 1


SONAR's first step is to cluster duplicate and similar sequences together,
categorize sequences (full-length, productive, non-productive, etc.), and
identify regions such as V(D)J segments and CDR3.


```{r}
tbl <- subset(counts_stats$sonar, select = -c(chain, antibody_lineage))
colnames(tbl)[match("nonproductive", colnames(tbl))] <- "nonprod"
tbl$GoodRatio <- tbl$good/(
  tbl$good + tbl$nonprod + tbl$indel + tbl$noCDR3 + tbl$noJ + tbl$noV + tbl$stop)
 drawtab(tbl, font_size=8)
```

To try to demonstrate that we have sufficient sequencing depth per cell and cell
counts per specimen, we can subsample the raw reads and tabulate number of
non-singleton clusters identified by SONAR.  If the number of clusters observed 
approaches the asymptote well before all reads are included, and if the
asymptote falls well below the number of cells (i.e., there are far fewer
sequences than cells) we have some confidence that we've sampled deeply.

```{r results="asis"}

plot_rarefaction_curve <- function(data, metadata, samp) {
  plot(
    x = data$ReadsUsed,
    y = data$UniqueCentroids / subset(metadata$samples, Sample == samp)$SpecimenCellCount,
    pch=19, cex=0.7,
    xlab="Reads Included", ylab="Clusters to Input Cells")
}

with(NULL, {

  for (idx in 1:nrow(metadata$antibody_lineages)) {
    lineage <- metadata$antibody_lineages$AntibodyLineage[idx]
    subject <- metadata$antibody_lineages$Subject[idx]
    samps <- subset(metadata$samples, grepl("IgG+", SpecimenCellType) & SpecimenSubject == subject)
    if (nrow(samps) == 0) {
      next
    }
    samps$Lineage <- lineage
    fps <- file.path(
      ROOT,
      "analysis/reporting",
      paste(samps$Specimen, samps$Lineage, samps$Chain, samps$Type, sep = "."),
      "sonar_clusters_rarefaction.csv")
      for (s_idx in 1:nrow(samps)) {
        cat(paste0(
          "\n\n### ",
          "Rarefaction Curve - ",
          samps$Specimen[s_idx], ".", samps$Lineage[s_idx], ".", samps$Chain[s_idx], ".", samps$Type[s_idx]),
          "\n\n")
        fp <- fps[s_idx]
        if (file.exists(fp)) {
          data <- read.csv(fp)
          # TODO make y-axis scaled by cell count, x axis by total reads
          plot_rarefaction_curve(data, metadata, samps$Sample[s_idx])
        } else {
        cat("No rarefaction curve available.\n")
      }
    }
  }
})
```