# Technical Details and Workflow

Raw read data goes in `data/{run}/`, either by bcl2fastq from original sequencer
output or downlaoded according to `metadata/runs.csv`.  This is controlled by
`data.smk`.

## Processing by Sample

The reads are chunked into a fixed number of files in `analysis/data/{run}` for
easy parallel processing, and then demultiplexed into
`analysis/demux/{run}/{chunk}/` according to `metadata/samples.csv`, controlled
by `by-run/demux.smk`.  Likewise for `analysis/trim/` in chunks, via
`by-run/trim.smk`, and finally the chunks are merged again into
`analysis/reads-by-sample` via `by-run/reads.smk`.

At this point reads are processed and grouped by sequencer sample.

## Processing by Specimen

Reads by sequencer sample are combined to a per-specimen/per-chain level in
`analysis/reads-by-specimen/{chain}.{chain_type}/{specimen}` files for use in
pRESTO and IgDiscover, via `by-specimen/by-specimen.smk`.

pRESTO (via `by-specimen/presto.smk`) writes to
`analysis/presto/{task}/{chain}.{chain_type}/{specimen}` where `task` is data,
assemble, qual, etc.  IgDiscover pulls from `analysis/presto/data` although this
is actually just an uncompressed version of each by-specimen file so it can
probably be reorganized.

IgDiscover writes to `analysis/igdiscover/{chain}.{chain_type}/{specimen}` via
`by-specimen/igdiscover.smk`.
 
## Processing by Subject

SONAR (via `by-subject/sonar.smk`) uses IgDiscover's output to define the
germline database for each subject in
`analysis/sonar/{subject}/{chain}.{chain_type}/germline.{segment}.fasta`
and analyzes specimens on a per-lineage basis in
`analysis/sonar/{subject}/{chain}.{chain_type}/{antibody_lineage}/`. 
`.../{chain}.{chain_type}/mab.fasta`, created from
`metadata/antibody_isolates.csv`, defines the set of mature antibody sequences
per subject per chain.  The input file `.../{antibody_lineage}/{specimen}.fastq`
is taken from pRESTO's FWD_primers-pass file.

SONAR's modules 1 and 2 run on a per-specimen basis, but module 3 combines
information from specimens into `.../{antibody_lineage}/longitudinal/`.

## Reporting

`analysis/counts` contains files referencing other files in `analysis/`, giving
read counts appropriately per file type, via `counts.smk`.  These are used to
build summary tables for the report.